From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrIvanPlays <ivan@mrivanplays.com>
Date: Wed, 15 Jul 2020 21:25:15 +0300
Subject: [PATCH] Origami config


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 6e01bf2d52e8bb6de7395f50c12f16c64aef72ae..a82cbf6c054113a6d6144b3629cd19f04d2199b0 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1447,6 +1447,14 @@ public interface Server extends PluginMessageRecipient {
         }
         // Paper end
 
+        // Origami start - add config to timings report
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getOrigamiConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Origami end
+
         /**
          * Sends the component to the player
          *
