From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrIvanPlays <ivan@mrivanplays.com>
Date: Wed, 15 Jul 2020 21:25:15 +0300
Subject: [PATCH] Origami config


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 457007005f9ff5f6818c559635379a2e698fc4d7..961bdd80d6074291361186624447e6ea9f83f4ad 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -1387,6 +1387,14 @@ public interface Server extends PluginMessageRecipient {
         }
         // Paper end
 
+        // Origami start - add config to timings report
+        @NotNull
+        public org.bukkit.configuration.file.YamlConfiguration getOrigamiConfig()
+        {
+            throw new UnsupportedOperationException("Not supported yet.");
+        }
+        // Origami end
+
         /**
          * Sends the component to the player
          *
