From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Spottedleaf <spottedleaf@spottedleaf.dev>
Date: Sat, 8 Aug 2020 18:12:16 +0300
Subject: [PATCH] Fix large move vectors killing the server and reset the
 CachedList on tick


diff --git a/src/main/java/com/mrivanplays/origami/CachedLists.java b/src/main/java/com/mrivanplays/origami/CachedLists.java
index 3e4538260ba0cba5443a3556481a5e52a6abc148..b6fb8e2f4905cfeba63a886ff2ce5f4597646717 100644
--- a/src/main/java/com/mrivanplays/origami/CachedLists.java
+++ b/src/main/java/com/mrivanplays/origami/CachedLists.java
@@ -27,4 +27,8 @@ public class CachedLists {
         ((UnsafeList) entityList).setSize(0);
         tempGetEntitiesListInUse = false;
     }
+
+    public static void reset() {
+        TEMP_GET_ENTITIES_LIST.completeReset();
+    }
 }
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index f92b8dc02cd32190233cd3c0e818714c1302a010..6a68e3a3bb33815f0092d4f0d4c31289a3c18526 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -1264,6 +1264,8 @@ public abstract class MinecraftServer extends IAsyncTaskHandlerReentrant<TickTas
         }
         // Paper end
 
+        com.mrivanplays.origami.CachedLists.reset(); // Origami
+
         // Paper start
         long endTime = System.nanoTime();
         long remaining = (TICK_TIME - (endTime - lastTick)) - catchupTime;
diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 3074a136568f3a0c937c252ad1aaf9edf2b4b7d5..488651f2e4d02f5ceed684bd4f77494c16fc6063 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -338,19 +338,24 @@ public class PlayerConnection implements PacketListenerPlayIn {
 
             if (entity != this.player && entity.getRidingPassenger() == this.player && entity == this.r) {
                 WorldServer worldserver = this.player.getWorldServer();
-                double d0 = entity.locX();
-                double d1 = entity.locY();
-                double d2 = entity.locZ();
-                double d3 = packetplayinvehiclemove.getX();
-                double d4 = packetplayinvehiclemove.getY();
-                double d5 = packetplayinvehiclemove.getZ();
+                double d0 = entity.locX(); double fromX = d0; // Origami - OBFHELPER
+                double d1 = entity.locY(); double fromY = d1; // Origami - OBFHELPER
+                double d2 = entity.locZ(); double fromZ = d2; // Origami - OBFHELPER
+                double d3 = packetplayinvehiclemove.getX(); double toX = d3; // Origami - OBFHELPER
+                double d4 = packetplayinvehiclemove.getY(); double toY = d4; // Origami - OBFHELPER
+                double d5 = packetplayinvehiclemove.getZ(); double toZ = d5; // Origami - OBFHELPER
                 float f = packetplayinvehiclemove.getYaw();
                 float f1 = packetplayinvehiclemove.getPitch();
                 double d6 = d3 - this.s;
                 double d7 = d4 - this.t;
                 double d8 = d5 - this.u;
                 double d9 = entity.getMot().g();
-                double d10 = d6 * d6 + d7 * d7 + d8 * d8;
+                // Origami start
+                double currDeltaX = toX - fromX;
+                double currDeltaY = toY - fromY;
+                double currDeltaZ = toZ - fromZ;
+                double d10 = Math.max(d6 * d6 + d7 * d7 + d8 * d8, (currDeltaX * currDeltaX + currDeltaY * currDeltaY + currDeltaZ * currDeltaZ) - 1);
+                // Origami end
 
 
                 // CraftBukkit start - handle custom speeds and skipped ticks
